### PR TITLE
Try to fix out of view statements

### PR DESCRIPTION
**Pull request by vstakhov**
This issue happens when some peer sends a good but already known `Seconded` statement and the statement-distribution code does not update the `statements_received` field in the `peer_knowledge` structure. Subsequently, a `Valid` statement causes `out-of-view` message that is incorrectly emitted and causes reputation lose.

### CODE REVIEW

File: node/network/statement-distribution/src/lib.rs [hunk 1]
```diff
@@ -1459,6 +1459,13 @@ async fn handle_incoming_message<'a>(
 			Err(DeniedStatement::NotUseful) => return None,
 			Err(DeniedStatement::UsefulButKnown) => {
 				report_peer(ctx, peer, BENEFIT_VALID_STATEMENT).await;
+				// Note a received statement in the peer data
+				// TODO: Understand if we need to share our active statements with the
+				// peer in this case.
+				peer_data
+					.receive(&relay_parent, &fingerprint, max_message_count)
+					.unwrap_or_else(|_| unreachable!("checked in `check_can_receive` above; qed"));
```
Comment (drahnr) [inline node/network/statement-distribution/src/lib.rs:1467]:
```suggestion
					.expect!("checked in `check_can_receive` above; qed");
```

File: node/network/statement-distribution/src/tests.rs [hunk 2]
```diff
@@ -1842,6 +1842,324 @@ fn peer_cant_flood_with_large_statements() {
...
+// Secondly, the proposed change to update received knowledge of a peer in the
+// function `handle_incoming_message` must also be removed to reproduce the issue.
+// This is done after `active_head.check_useful_or_unknown` check where a statement
+// is useful but known.
+#[test]
+fn handle_multiple_seconded_statements() {
+	let hash_a = Hash::repeat_byte(1);
+
+	let candidate = {
+		let mut c = dummy_committed_candidate_receipt(dummy_hash());
```
Comment (drahnr) [inline node/network/statement-distribution/src/tests.rs:1862]: shouldn't this be constructed with `hash_a`?

File: node/network/statement-distribution/src/tests.rs [hunk 3]
```diff
@@ -1842,6 +1842,324 @@ fn peer_cant_flood_with_large_statements() {
...
+// It could be done, for example, by avoiding calling to `peer_knowledge.send`:
+// `let peers_to_send: Vec<(PeerId, bool)> = peers_to_send` <--- here
+//
+// Secondly, the proposed change to update received knowledge of a peer in the
+// function `handle_incoming_message` must also be removed to reproduce the issue.
+// This is done after `active_head.check_useful_or_unknown` check where a statement
+// is useful but known.
+#[test]
+fn handle_multiple_seconded_statements() {
+	let hash_a = Hash::repeat_byte(1);
```
Comment (drahnr) [inline node/network/statement-distribution/src/tests.rs:1859]: nit: make it a bit more concise to use this throughout

```suggestion
	let relay_parent = Hash::repeat_byte(1);
```

File: node/network/statement-distribution/src/tests.rs [hunk 4]
```diff
@@ -1842,6 +1842,324 @@ fn peer_cant_flood_with_large_statements() {
...
+	let candidate = {
+		let mut c = dummy_committed_candidate_receipt(hash_a);
+		c.descriptor.relay_parent = hash_a;
+		c.descriptor.para_id = 1.into();
+		c
+	};
+	let candidate_hash = candidate.hash();
+
+	let peer_a = PeerId::random();
+	let peer_b = PeerId::random();
```
Comment (drahnr) [inline node/network/statement-distribution/src/tests.rs:1870]:
```suggestion
	let peer_b = PeerId::random();
	assert_ne!(peer_a, peer_b);
```

File: node/network/statement-distribution/src/tests.rs [hunk 5]
```diff
@@ -1842,6 +1842,324 @@ fn peer_cant_flood_with_large_statements() {
...
+// function `handle_incoming_message` must also be removed to reproduce the issue.
+// This is done after `active_head.check_useful_or_unknown` check where a statement
+// is useful but known.
+#[test]
+fn handle_multiple_seconded_statements() {
+	let hash_a = Hash::repeat_byte(1);
+
+	let candidate = {
+		let mut c = dummy_committed_candidate_receipt(hash_a);
+		c.descriptor.relay_parent = hash_a;
```
Comment (drahnr) [inline node/network/statement-distribution/src/tests.rs:1863]:
```suggestion
```

File: node/network/statement-distribution/src/tests.rs [hunk 6]
```diff
@@ -1842,6 +1842,324 @@ fn peer_cant_flood_with_large_statements() {
...
+		Sr25519Keyring::Alice.pair(),
+		Sr25519Keyring::Bob.pair(),
+		Sr25519Keyring::Charlie.pair(),
+	];
+
+	let session_info = make_session_info(validators, vec![]);
+
+	let session_index = 1;
+
+	let pool = sp_core::testing::TaskExecutor::new();
```
Comment (drahnr) [inline node/network/statement-distribution/src/tests.rs:1882]: Most of these test were written before we had the concept of a test harness and the virtual overseer, so this is lots of manual work we eventually want to formalize. This is clearly out of scope for this PR, but we should keep that in mind.

File: node/network/statement-distribution/src/tests.rs [hunk 7]
```diff
@@ -1842,6 +1842,324 @@ fn peer_cant_flood_with_large_statements() {
...
+	let (ctx, mut handle) = polkadot_node_subsystem_test_helpers::make_subsystem_context(pool);
+
+	let (statement_req_receiver, _) = IncomingRequest::get_config_receiver();
+
+	let bg = async move {
```
Comment (drahnr) [inline node/network/statement-distribution/src/tests.rs:1887]:
```suggestion
	let virtual_overseer = async move {
```

File: node/network/statement-distribution/src/tests.rs [hunk 8]
```diff
@@ -1842,6 +1842,324 @@ fn peer_cant_flood_with_large_statements() {
...
+				msg: StatementDistributionMessage::NetworkBridgeUpdateV1(
+					NetworkBridgeEvent::PeerMessage(
+						peer_b.clone(),
+						protocol_v1::StatementDistributionMessage::Statement(
+							hash_a,
+							statement.clone().into(),
+						),
+					),
+				),
+			})
```
Comment (drahnr) [inline node/network/statement-distribution/src/tests.rs:2049]:
```suggestion
			fn from_overseer(handle: &OverseerHandle, msg: impl Into<StatementDistributionMessage>) {
			handle.send(FromOverseer::Communication { msg: msg.into() })
			}
			
			from_overseer(&mut handle, StatementDistributionMessage::NetworkBridgeUpdateV1(
					NetworkBridgeEvent::PeerMessage(
						peer_b.clone(),
						protocol_v1::StatementDistributionMessage::Statement(
							hash_a,
							statement.clone().into(),
						),
					),
				);
```

File: node/network/statement-distribution/src/tests.rs [hunk 9]
```diff
@@ -1842,6 +1842,324 @@ fn peer_cant_flood_with_large_statements() {
 	executor::block_on(future::join(test_fut, bg));
 }
 
+// This test addresses an issue when received knowledge is not updated on a
```
Comment (drahnr) [inline node/network/statement-distribution/src/tests.rs:1845]: We should elaborate the issue observed here, `an issue` will not be enough in 3 months time.

File: node/network/statement-distribution/src/tests.rs (hunk 5 above)
Comment (vstakhov) [reply node/network/statement-distribution/src/tests.rs:1863]: Actually, the second statement is also not needed as `dummy_committed_candidate_receipt` handles it.

File: node/network/statement-distribution/src/tests.rs (hunk 9 above)
Comment (drahnr) [reply node/network/statement-distribution/src/tests.rs:1845]: We should create one, I am not aware of one at this time.

**Review by drahnr — CHANGES_REQUESTED** [review]
Soundness of the fix checks out.

The test case itself needs attention.

Since modifying the subsystem in it's entirety is not feasible, there are a few things to do:

* add modification functions guarded behind `#[cfg(test)]` to modify the internal state as needed
* modify `util::choose_random_subset` to consume an `impl Rng` such that in test cases this can be modified as desired

With these changes in place the leading head comment of the test case can be removed in favor of the to be created issue link besides a few more notes inline, where the state is modified with the yet to be created functions.

An alternative approach would be, to not instantiate the full subsystem but only test the incoming messages, and update the view explicitly. The state is too complex though to do so imho.

File: node/network/statement-distribution/src/tests.rs (hunk 9 above)
Comment (vstakhov) [reply node/network/statement-distribution/src/tests.rs:1853]: I have linked this PR for reference.

File: node/network/statement-distribution/src/lib.rs [hunk 10]
```diff
@@ -61,7 +61,6 @@ use util::runtime::RuntimeInfo;
 use std::collections::{hash_map::Entry, HashMap, HashSet, VecDeque};
 
 use fatality::Nested;
-
```
Comment (drahnr) [inline node/network/statement-distribution/src/lib.rs:64]: nit: I assume this was unintentional?

File: node/network/statement-distribution/src/lib.rs [hunk 11]
```diff
@@ -115,16 +114,19 @@ const LOG_TARGET: &str = "parachain::statement-distribution";
 const MAX_LARGE_STATEMENTS_PER_SENDER: usize = 20;
 
 /// The statement distribution subsystem.
-pub struct StatementDistributionSubsystem {
+pub struct StatementDistributionSubsystem<R: rand::Rng> {
```
Comment (drahnr) [inline node/network/statement-distribution/src/lib.rs:117]:
```suggestion
pub struct StatementDistributionSubsystem<R> {
```

Comment (drahnr) [reply node/network/statement-distribution/src/lib.rs:117]: Usually bounds are avoided in struct generics, unless they can't be expressed as part of the body.

File: node/subsystem-util/src/lib.rs [hunk 12]
```diff
@@ -276,33 +277,41 @@ pub fn find_validator_group(
 
 /// Choose a random subset of `min` elements.
 /// But always include `is_priority` elements.
-pub fn choose_random_subset<T, F: FnMut(&T) -> bool>(
+pub fn choose_random_subset<T, F: FnMut(&T) -> bool>(is_priority: F, v: &mut Vec<T>, min: usize) {
```
Comment (drahnr) [inline node/subsystem-util/src/lib.rs:280]: Unless there is a good and proven perf improvement, I am not a big fan of the C-style paradigm, pass a ptr or ref and modify in that fn, I'd rather consume, modify and return.

Comment (vstakhov) [reply node/subsystem-util/src/lib.rs:280]: Well, it seems that this paradigm is more descriptive than taking `mut Vec` and it is used in many other places in Rust standard library (e.g. `sort` or even `shuffle` used in this function). I was also unsure about how to preserve `mut` in the forwarded chains (e.g. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6020634a0415e999e0e162f3f4f777a2). With `&mut Vec` this function does exactly what is expected: get a vector and modifies it in place.

**Review by tdimitrov — APPROVED** [review]
Good catch @vstakhov!
The change makes sense to me after following your explanation and commits but doesn't look trivial to catch.

**Comment by vstakhov** [comment]
Bot merge.

**Comment by vstakhov** [comment]
bot merge

**Pull request merged**